nvflare/job_config/base_app_config.py

```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os.path
+from abc import ABC
+from typing import Dict, List
+
+from nvflare.apis.filter import Filter
+from nvflare.apis.fl_component import FLComponent
+
+
+class BaseAppConfig(ABC):
+    """BaseAppConfig holds the base essential component data for the ServerApp and ClientApp, including the
+    task_data_filters, task_result_filters, system components and used external scripts.
+
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self.task_data_filters: [(List[str], Filter)] = []
+        self.task_result_filters: [(List[str], Filter)] = []
+        self.components: Dict[str, object] = {}
+        self.ext_scripts = []
+
+        self.handlers: [FLComponent] = []
+
+    def add_component(self, cid: str, component):
+        if cid in self.components.keys():
+            raise RuntimeError(f"Component with ID:{cid} already exist.")
+
+        self.components[cid] = component
+
+        if isinstance(component, FLComponent):
+            self.handlers.append(component)
+
+    def add_task_data_filter(self, tasks: List[str], filter: Filter):
+        self._add_task_filter(tasks, filter, self.task_data_filters)
+
+    def add_task_result_filter(self, tasks: List[str], filter: Filter):
+        self._add_task_filter(tasks, filter, self.task_result_filters)
+
+    def add_ext_script(self, ext_script: str):
```

Comment:
Does this belong to BaseApp? Or it is a special ClientApp?

Reply:
This is for the  BaseApp. ServerApp could have it as well.